File: apps/frontend/lib/api/mock-interview.ts
import { apiFetch, apiPost } from './client';

export interface InterviewOption {
  option_id: string;
  text: string;
}

export interface PublicInterviewQuestion {
  question_id: string;
  topic: string;
  difficulty: number;
  stem: string;
  options: InterviewOption[];
  source_requirement?: string | null;
}

export interface InterviewStats {
  total_answered: number;
  correct_answers: number;
  accuracy: number;
  average_response_time_ms: number;
  mastery_score: number;
  forgetting_score: number;
  current_difficulty: number;
  questions_remaining: number;
  predictor_confidence?: number | null;
  predictor_source?: string | null;
}

export interface InterviewHistoryItem {
  question_id: string;
  topic: string;
  difficulty: number;
  correct: boolean;
  response_time_ms: number;
  answered_at: string;
}

export interface MockInterviewSessionPayload {
  session_id: string;
  opening_message?: string | null;
  context: {
    core_requirements?: string[];
    matched_skills?: string[];
    missing_skills?: string[];
    responsibilities?: string[];
    resume_highlights?: string[];
    focus_areas?: string[];
    match_ratio?: number;
  };
  stats: InterviewStats;
  current_question?: PublicInterviewQuestion | null;
  history?: InterviewHistoryItem[];
}

export interface MockInterviewAnswerResponse {
  session_id: string;
  question_id: string;
  correct: boolean;
  correct_option_id: string;
  explanation: string;
  answer_summary: string;
  completed: boolean;
  stats: InterviewStats;
  next_question?: PublicInterviewQuestion | null;
}

export async function createMockInterviewSession(payload: {
  resume_id: string;
  job_id?: string | null;
  language: string;
  question_count: number;
}): Promise<MockInterviewSessionPayload> {
  const res = await apiPost('/mock-interview/sessions', payload);
  if (!res.ok) {
    const text = await res.text().catch(() => '');
    throw new Error(text || `Failed to create mock interview session (status ${res.status}).`);
  }
  return (await res.json()) as MockInterviewSessionPayload;
}

export async function submitMockInterviewAnswer(
  sessionId: string,
  payload: {
    question_id: string;
    selected_option_id: string;
    response_time_ms: number;
  }
): Promise<MockInterviewAnswerResponse> {
  const res = await apiPost(
    `/mock-interview/sessions/${encodeURIComponent(sessionId)}/answers`,
    payload
  );
  if (!res.ok) {
    const text = await res.text().catch(() => '');
    throw new Error(text || `Failed to submit mock interview answer (status ${res.status}).`);
  }
  return (await res.json()) as MockInterviewAnswerResponse;
}

export async function fetchMockInterviewSession(
  sessionId: string
): Promise<MockInterviewSessionPayload> {
  const res = await apiFetch(`/mock-interview/sessions/${encodeURIComponent(sessionId)}`);
  if (!res.ok) {
    const text = await res.text().catch(() => '');
    throw new Error(text || `Failed to load mock interview session (status ${res.status}).`);
  }
  return (await res.json()) as MockInterviewSessionPayload;
}

File: apps/backend/app/schemas/mock_interview.py
"""Schemas for adaptive mock interview."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field, field_validator


class InterviewOption(BaseModel):
    """Single multiple-choice option."""

    option_id: str
    text: str


class PublicInterviewQuestion(BaseModel):
    """Question payload safe to send before the user answers."""

    question_id: str
    topic: str
    difficulty: int = Field(ge=1, le=5)
    stem: str
    options: list[InterviewOption]
    source_requirement: str | None = None


class InterviewStats(BaseModel):
    """Aggregate stats for the current interview session."""

    total_answered: int = 0
    correct_answers: int = 0
    accuracy: float = 0.0
    average_response_time_ms: float = 0.0
    mastery_score: float = 0.0
    forgetting_score: float = 0.0
    current_difficulty: int = Field(ge=1, le=5)
    questions_remaining: int = 0
    predictor_confidence: float | None = None
    predictor_source: str | None = None


class InterviewHistoryItem(BaseModel):
    """Completed answer history entry."""

    question_id: str
    topic: str
    difficulty: int = Field(ge=1, le=5)
    correct: bool
    response_time_ms: int = Field(ge=0)
    answered_at: str


class MockInterviewSessionCreateRequest(BaseModel):
    """Request to start a mock interview session."""

    resume_id: str
    job_id: str | None = None
    language: str = Field(default="en", min_length=2, max_length=16)
    question_count: int = Field(default=5, ge=1, le=15)


class MockInterviewSessionCreateResponse(BaseModel):
    """Response returned when a session starts."""

    session_id: str
    opening_message: str
    context: dict[str, Any]
    stats: InterviewStats
    current_question: PublicInterviewQuestion


class MockInterviewAnswerRequest(BaseModel):
    """Answer submission payload."""

    question_id: str
    selected_option_id: str = Field(min_length=1, max_length=8)
    response_time_ms: int = Field(ge=0, le=600000)

    @field_validator("selected_option_id")
    @classmethod
    def normalize_option_id(cls, value: str) -> str:
        return value.strip().upper()


class MockInterviewAnswerResponse(BaseModel):
    """Response after evaluating one answer."""

    session_id: str
    question_id: str
    correct: bool
    correct_option_id: str
    explanation: str
    answer_summary: str
    completed: bool
    stats: InterviewStats
    next_question: PublicInterviewQuestion | None = None


class MockInterviewSessionResponse(BaseModel):
    """Session status payload for refresh and recovery."""

    session_id: str
    opening_message: str | None = None
    context: dict[str, Any]
    stats: InterviewStats
    current_question: PublicInterviewQuestion | None = None
    history: list[InterviewHistoryItem] = Field(default_factory=list)
